test-data/unit/check-union-or-syntax.test

```diff
@@ -196,3 +196,16 @@ def f(x: Union[int, str, None]) -> None:
     else:
         reveal_type(x)  # N: Revealed type is "None"
 [builtins fixtures/isinstance.pyi]
+
+[case testUnionOrSyntaxMetaclassOverride]
+# flags: --python-version 3.10
+class U(type): pass
+
+class Meta(type):
+    def __or__(self, other) -> U: pass # type: ignore
```

Comment:
Let's also test regular metaclass with `def __or__(self: T, other: U) -> Union[T, U]:`

Reply:
🤔 Maybe both?

Reply:
I think the test you suggest is a different issue tbh. The revealed type for unions for this is all over the place in master (so this is not considering my changes, which don't have any affect on this behavior):
```
class A: ...
class B: ...
reveal_type(A | B) # "types.UnionType"
C = A | B
reveal_type(C) # "builtins.object"
```

```
from typing import TypeVar

T = TypeVar('T')
U = TypeVar('U')

class Meta(type):
    def __or__(self: T, other: U) -> Union[T, U]: pass # type: ignore

class F(metaclass=Meta): pass

reveal_type(F | F)  # "Any"
UF = F | F
reveal_type(UF)  # "builtins.object"
```

If you do the same as above but add `from typing import Union`
```
reveal_type(F | F)  # "def () -> test2.F"
UF = F | F
reveal_type(UF)  # "builtins.object"
```

As a side note, in my branch there is now no difference between `reveal_type(F | F)` and `reveal_type(UF)` in any of these cases (they all follow the first result). So it's at least a bit more consistent there.